docs/design/2018-09-04-histograms-in-plan.md

```diff
@@ -0,0 +1,115 @@
+
+- Semi join: It’s something similar to inner join. But no data expanding occurs. When we maintain the range information. We can get a nearly accurate answer of its row count.
+
+- Anti semi join: Same with semi join.
+
+For `Selection`, just use it to calculate the selectivity. 
+
+For `DataSource`, if it’s a non-partitioned table, we just maintain the map. If it’s a partitioned table, we now only store the statistics of each partition So we need to merge them. We’ll need a cache or something else to ensure that we won’t merge them each time we need it, which will consume tooooo much time and memory space.
+
+For other plan operators or `DataSource` which do not contain any histogram, we just use the original way to maintain `statsInfo`. We won’t maintain histograms for them. Taking `Union` for an example, if we maintain histogram for it, we need to merge the histograms which is really an expensive operation. But we can consider maintaining this if a hint is given.
+
+When we maintain `rangesOfXXX` or `max and min values`:
+
+Most things are the same with above. But there will be something more to do.
+
+For `Selection` and `Join`, we need to cut off the things which are not in ranges when doing estimation. And update the ranges information after the estimation. Also the NDV of column can be estimated more accurately.
+
+For `Aggregation`, we only need to cut off the things which are not in ranges when doing estimation. There is no need to update the ranges information.
+
+For `TopN`, we now have alibity to maintain histograms of the order-by items.
+
+
+## Rationale
+
+### How other systems solve the same issue?
+
+I’ve looked into Spark. They did nearly the same thing with what I said. They only maintain the max and min values, no `ranges` information. And they don’t have index, so they only maintain the column’s max/min value which make problem far more much easier.
+
+As for Orca and Calcite, I haven’t discovered where they maintain this information. But there’s something about statistics in Orca’s paper. According to the paper, I think they construct new histogram during planning and cache it for not building to often.
+
+### What is the disadvantage of this design?
+
+This may have side effects on OLTP performance. But I’ll try to reduce it.
+ 
+For now, only join reorder and the position `after logical optimize before physical optimize` will trigger this mechanism and it’s controlled by `tidb_optimizer_selectivity_level` which is disabled by default. This may not have much more side effects on simple point queries.
+
+And the `expectedCount` we used in physical plan is something same with `Limit`. So the row count modification during physical plan won’t be affected.
+
+After we switch to the cascade like planner. The rule that needs cost to make decision is still a small set of all. And the existence of `Group` can also help us. If we lazily construct the `statsInfo`, this may not be the bottleneck.
```

Comment:
After we switch to the **cascade-like** planner. The rule that needs `cost` to make **a** decision is still a small set of all.